exchange/bitswap/bitswap.go

```diff
@@ -340,7 +323,23 @@ func (bs *Bitswap) HasBlock(blk blocks.Block) error {
 	return nil
 }
 
+// SessionsForBlock returns a slice of all sessions that may be interested in the given cid
+func (bs *Bitswap) SessionsForBlock(c *cid.Cid) []*Session {
+	bs.sessLk.Lock()
+	defer bs.sessLk.Unlock()
+
+	var out []*Session
+	for _, s := range bs.sessions {
+		if s.interestedIn(c) {
+			out = append(out, s)
+		}
+	}
+	return out
+}
+
 func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg.BitSwapMessage) {
+	atomic.AddUint64(&bs.messagesRecvd, 1)
```

Comment:
Atomic will only work if address is 8 bytes, aliased. This might not be the case if there are other variables in the structure, I suggest allocating separate structure for all metrics and keeping all variable that will be used atomically as `(u)int64` and at the front.